From e7b7c1648474b22d3f49422d5f818a5b03b208ff Mon Sep 17 00:00:00 2001
From: Naveen Kumar <naveen1.kumar@intel.com>
Date: Thu, 2 Apr 2026 13:46:02 +0530
Subject: [PATCH] onscreen/native: Guard early swap bail behind render source
 check
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

It is completely normal for a new swap to arrive while the previous
frame's next_frame slot is still occupied waiting for the flip.

Earlier, the incoming frame would supersede the previous next_frame
via assign_next_frame()'s clear_superseded_frame() after resuming
from suspend.

However, now it is getting discarded before it even reaches
assign_next_frame().

v1: Remove comment as it already explains in the block below
(Michel Dänzer)

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/5003>
---
 src/backends/native/meta-onscreen-native.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/backends/native/meta-onscreen-native.c b/src/backends/native/meta-onscreen-native.c
index 62d041686ca..ac01c32fc10 100644
--- a/src/backends/native/meta-onscreen-native.c
+++ b/src/backends/native/meta-onscreen-native.c
@@ -1781,7 +1781,7 @@ meta_onscreen_native_swap_buffers_with_damage (CoglOnscreen    *onscreen,
   COGL_TRACE_BEGIN_SCOPED (MetaRendererNativeSwapBuffers,
                            "Meta::OnscreenNative::swap_buffers_with_damage()");
 
-  if (onscreen_native->next_frame != NULL)
+  if (onscreen_native->render_source && onscreen_native->next_frame != NULL)
     {
       /* This is needed to make sure we lock the correct front buffer for the
        * new frame, as eglSwapBuffers() swaps what is the active front buffer.
-- 
GitLab


